### PR TITLE
Prices refactor for perps v2

### PR DESCRIPTION
**Pull request by avclarke**
TODO:

- [x] Hook up latest price to trading view chart
- [x] Fix exchange trading synth to synth
- [x] Check correct prices are used in various places around futures components, e.g. skew adjusted, offchain, onchain

### CODE REVIEW

File: sdk/services/prices.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import KwentaSDK from 'sdk';
+
+import { MARKET_ASSETS_BY_PYTH_ID } from 'sdk/constants/futures';
+import { ADDITIONAL_SYNTHS, PYTH_IDS } from 'sdk/constants/prices';
+import { CurrencyRate, PricesMap, PriceType, SynthRatesTuple } from 'sdk/types/common';
+import { FuturesMarketKey } from 'sdk/types/futures';
+import { getPythNetworkUrl, normalizePythId } from 'sdk/utils/futures';
+import { startInterval } from 'sdk/utils/interval';
+import { scale } from 'utils/formatters/number';
+import { MarketAssetByKey } from 'utils/futures';
```
[the remark this answers is not in the document]
Comment (avclarke) [reply sdk/services/prices.ts:14]: Yea good call, forgot to move that one

File: sections/dashboard/FuturesMarketsTable/FuturesMarketsTable.tsx [hunk 2]
```diff
@@ -28,6 +33,7 @@ const FuturesMarketsTable: FC = () => {
 	const pastRates = useRecoilValue(pastRatesState);
 	const futuresVolumes = useAppSelector(selectMarketVolumes);
 	const accountType = useAppSelector(selectFuturesType);
+	const marketPrice = useAppSelector(selectLatestMarketPrice);
```
Comment (Tburm) [inline sections/dashboard/FuturesMarketsTable/FuturesMarketsTable.tsx:36]: This returns the price for the selected market, so duplicates the price here. I think it should be `selectedMarketPrices` and index by asset for each market.

image

File: sections/dashboard/FuturesPositionsTable/MobilePositionRow.tsx [hunk 3]
```diff
@@ -6,15 +6,19 @@ import ChangePercent from 'components/ChangePercent';
...
 import { getDisplayAsset, isDecimalFour } from 'utils/futures';
 
 type MobilePositionRowProps = {
+	// TODO: specify type
 	row: any;
 	onClick(): void;
 };
 
 const MobilePositionRow: React.FC<MobilePositionRowProps> = ({ row, onClick }) => {
+	const marketPrice = useAppSelector(selectMarketPrice);
```
Comment (Tburm) [inline sections/dashboard/FuturesPositionsTable/MobilePositionRow.tsx:21]: `selectMarketPrices` here as well, index by asset name

File: sections/dashboard/SynthBalancesTable/SynthBalancesTable.tsx [hunk 4]
```diff
@@ -60,14 +60,14 @@ const SynthBalancesTable: FC<SynthBalancesTableProps> = ({ exchangeTokens }) =>
...
-	const exchangeRates = useAppSelector(selectExchangeRates);
+	const prices = useAppSelector(selectPrices);
 	const { synthBalances } = useAppSelector(selectBalances);
 
 	const synthTokens = useMemo(() => {
 		return synthBalances.map((synthBalance: SynthBalance) => {
 			const { currencyKey, balance, usdBalance } = synthBalance;
 
-			const price = exchangeRates && exchangeRates[currencyKey];
+			const price = prices && prices[currencyKey].offChain;
```
Comment (Tburm) [inline sections/dashboard/SynthBalancesTable/SynthBalancesTable.tsx:70]: Should this be onchain as well? Spot balances would trade via atomic swaps, which don't use offchain oracle price.

File: sections/homepage/Assets/Assets.tsx [hunk 5]
```diff
@@ -208,10 +209,10 @@ const Assets = () => {
 					key: market.asset,
 					name: market.asset[0] === 's' ? market.asset.slice(1) : market.asset,
 					description: description.split(' ')[0],
-					price: market.price.toNumber(),
+					price: marketPrice.toNumber(),
```
Comment (Tburm) [inline sections/homepage/Assets/Assets.tsx:212]: Need to index by asset here:
image

File: state/futures/selectors.ts [hunk 6]
```diff
@@ -102,15 +96,44 @@ export const selectIsolatedPriceImpact = createSelector(
...
+);
+
+export const selectLatestMarketPrice = createSelector(
+	selectMarketAsset,
+	selectPrices,
+	(marketAsset, prices) => {
+		const price = prices[marketAsset];
+		return price?.offChain ?? price?.onChain ?? wei(0);
+	}
+);
```
Comment (Tburm) [inline state/futures/selectors.ts:121]: Should this include order type in the same way `selectSkewAdjustedPrice` does? When market/delayed orders are selected, I think we want the onchain value.

Comment (Tburm) [reply state/futures/selectors.ts:121]: Maybe when we disable everything but delayed offchain it doesn't matter, just noting here.

File: sections/futures/MarketDetails/useGetMarketData.ts [hunk 7]
```diff
@@ -31,6 +33,8 @@ const useGetMarketData = (mobile?: boolean) => {
 
 	const pastRates = useRecoilValue(pastRatesState);
 	const futuresVolumes = useAppSelector(selectMarketVolumes);
+	const marketPrices = useAppSelector(selectMarketPrices);
+	const marketPrice = useAppSelector(selectLatestMarketPrice);
```
Comment (Tburm) [inline sections/futures/MarketDetails/useGetMarketData.ts:37]: I would think we display the skew adjusted price here. This is what's displayed as the "mark price" of the market, which would be oracle price adjusted by skew.

File: state/futures/selectors.ts (hunk 6 above)
Comment (avclarke) [reply state/futures/selectors.ts:121]: This is just if you don't care what price, you just want the latest for display, either off chain as priority or fallback to on chain

**Review by Tburm — COMMENTED** [review]
Left some comments, one additional  note is that the `PositionCard` price doesn't update when the pyth price updates. I think related to not using the skew adjusted value.

File: state/futures/actions.ts [hunk 8]
```diff
@@ -281,15 +282,18 @@ export const fetchIsolatedMarginTradePreview = createAsyncThunk<
...
 		if (!marketInfo) throw new Error('No market info');
 		const leverageSide = selectLeverageSide(getState());
 		try {
 			const preview = await sdk.futures.getIsolatedTradePreview(
 				marketInfo?.market,
 				sizeDelta,
-				marketInfo?.priceOracle,
-				marketInfo?.price,
+				skewAdjustedPrice,
+				marketPrice,
```
Comment (Tburm) [inline state/futures/actions.ts:296]: Looks like these are backwards

```suggestion
				marketPrice,
				skewAdjustedPrice,
```

**Review by Tburm — APPROVED** [review]
Only issue I see is no prices on exchange, but it's like that on the base perps v2 branch. Maybe we just disable it.

**Pull request merged**